### ISSUE TITLE
Regarding training using Multiclass dataset

### ISSUE THREAD

**Issue by AishaJaved123** [Henryjiepanli/Uncertainty-aware-Network#17]
I was wondering if its possible to use the modules proposed in your study in order to train the whole network for multiclass dataset. Is it possible? If yes than do i have to change some parts of the code or just set the num_classes to the number of classes my dataset have?

Thank you

**Comment by tunahan-oguz** [comment]
Hello! The architecture's uncertainty definition is designed for binary scenarios (can be inferred from code and the paper) I recommend our multi-class extension of UA-Net to train it on multi-class scenarios. We propose an uncertainty definition that utilizes per pixel entropy calculation.
https://github.com/tunahan-oguz/UncertaintyAwareMultiClassSegmentation

**Comment by Henryjiepanli** [comment]
Thanks for your attention to our work. We would like to clarify that the original Uncertainty Ranking Algorithm (URA) was designed for binary classification scenarios (i.e., single-class segmentation tasks). However, if your application involves multi-class segmentation (i.e., number of classes > 2), we can certainly help you adapt the URA to suit that setting.In the multi-class case, we cannot use a fixed threshold like 0.5 to distinguish uncertainty levels, as is common in binary segmentation. Instead, we recommend using the entropy of the softmax probabilities as a more suitable measure of uncertainty. Entropy captures the overall confidence distribution across all classes and provides a natural way to quantify uncertainty in multi-class predictions. Based on this idea, we have modified the URA accordingly. The updated implementation has been included in the attached UANet.py file for your reference. If you have any questions about the changes or how to integrate them into your pipeline, please feel free to reach out. -----原始邮件-----
发件人:Aisha ***@***.***>
发送时间:2025-07-21 14:42:11 (星期一)
收件人: Henryjiepanli/Uncertainty-aware-Network ***@***.***>
抄送: Subscribed ***@***.***>
主题: [Henryjiepanli/Uncertainty-aware-Network] Regarding training using Multiclass dataset (Issue #17)


AishaJaved123 created an issue (Henryjiepanli/Uncertainty-aware-Network#17)

I was wondering if its possible to use the modules proposed in your study in order to train the whole network for multiclass dataset. Is it possible? If yes than do i have to change some parts of the code or just set the num_classes to the number of classes my dataset have?

Thank you

**Comment by AishaJaved123** [comment]
Thank you for your suggestions. I am not able to see the attached file.

**Comment by Henryjiepanli** [comment]
send me your email account， i will resend u

----- Original Message -----

**Comment by Henryjiepanli** [comment]
> Thank you for your suggestions. I am not able to see the attached file.

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class Uncertainty_Rank_Algorithm(nn.Module):
    def __init__(self, num_bins=5):
        super(Uncertainty_Rank_Algorithm, self).__init__()
        self.num_bins = num_bins
        self.eps = 1e-8

    def forward(self, logits):
        probs = F.softmax(logits, dim=1)
        entropy = -torch.sum(probs * torch.log(probs + self.eps), dim=1)
        H_max = torch.log(torch.tensor(probs.shape[1], dtype=torch.float32)).to(logits.device)
        norm_entropy = entropy / H_max

        rank_map = torch.zeros_like(norm_entropy)
        thresholds = torch.linspace(0, 1, self.num_bins + 1).to(logits.device)
        for i in range(self.num_bins):
            mask = (norm_entropy >= thresholds[i]) & (norm_entropy < thresholds[i + 1])
            rank_map[mask] = self.num_bins - i

        return rank_map.unsqueeze(1).detach()
```

```python
class Uncertainty_Aware_Fusion_Module(nn.Module):
    def __init__(self, high_channel, low_channel, out_channel, num_classes):
        super(Uncertainty_Aware_Fusion_Module, self).__init__()
        self.rank = Uncertainty_Rank_Algorithm()
        self.conv_high = nn.Conv2d(high_channel, out_channel, 3, 1, 1)
        self.conv_low = nn.Conv2d(low_channel, out_channel, 3, 1, 1)
        self.conv_fusion = nn.Conv2d(2 * out_channel, out_channel, 3, 1, 1)
        self.seg_out = nn.Conv2d(out_channel, num_classes, 1)

    def forward(self, feature_low, feature_high, logits):
        rank_high = self.rank(logits)
        rank_high = F.interpolate(rank_high, size=feature_high.shape[2:], mode='bilinear', align_corners=True)
        weighted_high = rank_high * feature_high
        fused_high = self.conv_high(weighted_high)
        fused_high_up = F.interpolate(fused_high, size=feature_low.shape[2:], mode='bilinear', align_corners=True)

        logits_low = F.interpolate(logits, size=feature_low.shape[2:], mode='bilinear', align_corners=True)
        rank_low = self.rank(logits_low)
        weighted_low = rank_low * feature_low
        fused_low = self.conv_low(weighted_low)

        fusion = torch.cat([fused_high_up, fused_low], dim=1)
        fusion = self.conv_fusion(fusion)
        seg = self.seg_out(fusion)

        return fusion, seg
```

**Comment by AishaJaved123** [comment]
Thank you for the code, I have one more question regarding "map = map[:,1,:,:].unsqueeze(1)" in the CGM module. Does this focus on one class right? If yes then it should also be update to focus on multiclass.

**Comment by Henryjiepanli** [comment]
> Thank you for the code, I have one more question regarding "map = map[:,1,:,:].unsqueeze(1)" in the CGM module. Does this focus on one class right? If yes then it should also be update to focus on multiclass.

In this paper, we just focus on the binary segmentation of building pixels. If you want to use CGM on multiclass task, maybe u need to adjust the idea of how to utilize the coarse information and revise the corresponding code.
